textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/GramBooleanQuery.java

```diff
@@ -33,9 +33,138 @@
 	
 	GramBooleanQuery(QueryOp operator, int gramLength) {
 		this.operator = operator;
-		operandList = new ArrayList<String>();
-		subQueryList = new ArrayList<GramBooleanQuery>();
-		this.gramLength = gramLength;
+		operandSet = new HashSet<String>();
+		subQuerySet = new HashSet<GramBooleanQuery>();
+		GramBooleanQuery.gramLength = gramLength;
+	}
+	
+	
+	/**
+	 * This method takes a list of strings and adds them to the query tree. <br>
+	 * For example, if the list is {abcd, wxyz}, then: <br>
+	 * trigrams({abcd, wxyz}) = trigrams(abcd) OR trigrams(wxyz) <br>
+	 * OR operator is assumed for a list of strings. <br>
+	 * @param list, a list of strings to be added into query.
+	 */
+	void add(List<String> list) {
+		addOrNode(list);
+	}
+	
+	private void addOrNode(List<String> literalList) {
+		if (TranslatorUtils.minLenOfString(literalList) < 3) {
+			return;
+		}
+		if (literalList.size() == 0) {
+			return;
+		} else if (literalList.size() == 1) {
+			this.addAndNode(literalList.get(0));
+		} else {
+			GramBooleanQuery query = new GramBooleanQuery(QueryOp.OR);
+			for (String literal : literalList) {
+				query.addAndNode(literal);
+			}
+			this.subQuerySet.add(query);
+		}
+	}
+	
+	/**
+	 * This method takes a single string and adds it to the query tree. <br>
+	 * The string is converted to multiple n-grams with an AND operator. <br>
+	 * For example: if the string is abcd, then: <br>
+	 * trigrams(abcd) = abc AND bcd <br>
+	 * AND operator is assumed for a single string. <br>
+	 * @param literal
+	 */
+	private void addAndNode(String literal) {
+		if (literal.length() < gramLength) {
+			return;
+		} else if (literal.length() == gramLength) {
```

Comment:
I see.  It's a consequence of this design.  (The other design we discussed earlier won't have this problem, since all nodes are treated "equally.")

It will be good to add a comment to the code to explain why we are doing this (line 81).